openwisp_firmware_upgrader/admin.py

```diff
@@ -200,22 +201,59 @@ def __get_rate(self, value):
     aborted_rate.short_description = _('abortion rate')
 
 
+class DeviceFirmwareForm(forms.ModelForm):
+    class Meta:
+        model = load_model('DeviceFirmware')
+        fields = '__all__'
+
+    def _get_compatible_boards(self, instance):
+        # instance.device.model in instance.image.boards
+        # boards = FIRMWARE_IMAGE_MAP[instance.image.type]['boards']
+        device_model = instance.device.model
+        qs = load_model('FirmwareImage').objects.filter(
+            build__category__organization=instance.device.organization
+        )  # Or filter by type=...
+        list_of_ids = [
+            fw_image.pk for fw_image in qs if device_model in fw_image.boards
+        ]
+        return qs.filter(id__in=list_of_ids).order_by('-created')
+
+    def __init__(self, *args, **kwargs):
+        super().__init__(*args, **kwargs)
+        if 'instance' in kwargs:
+            instance = kwargs['instance']
+            self.fields['image'].queryset = self._get_compatible_boards(instance)
+
+
 class DeviceFirmwareInline(MultitenantAdminMixin, admin.StackedInline):
     model = load_model('DeviceFirmware')
+    form = DeviceFirmwareForm
     exclude = ['created']
+    select_related = ['device', 'image']
     readonly_fields = ['installed', 'modified']
     verbose_name = _('Device Firmware')
     verbose_name_plural = verbose_name
     extra = 0
     multitenant_shared_relations = ['device']
 
 
+class DeviceUpgradeOperationInline(UpgradeOperationInline):
+
+    verbose_name = _('In progress upgrades')
+    verbose_name_plural = verbose_name
+
+    def get_queryset(self, request):
+        qs = super().get_queryset(request)
+        return qs.filter(status='in-progress')
```

Comment:
I thought we said to show the last, so that it doesn't disappear when it's finished.
Ensure it's ordered by creation date (most recent first).

Reply:
Update: I think we can make it show only recent upgrades, so upgrades that have been created within something like a week